src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Documents/EditorDocumentManagerListener.cs

```diff
@@ -61,56 +78,126 @@ public override void Initialize(ProjectSnapshotManagerBase projectManager)
         }
 
         // Internal for testing.
-        internal void ProjectManager_Changed(object sender, ProjectChangeEventArgs e)
+#pragma warning disable VSTHRD100 // Avoid async void methods
+        internal async void ProjectManager_Changed(object sender, ProjectChangeEventArgs e)
+#pragma warning restore VSTHRD100 // Avoid async void methods
         {
-            switch (e.Kind)
+            try
             {
-                case ProjectChangeKind.DocumentAdded:
-                    {
-                        var key = new DocumentKey(e.ProjectFilePath, e.DocumentFilePath);
-                        var document = _documentManager.GetOrCreateDocument(key, _onChangedOnDisk, _onChangedInEditor, _onOpened, _onClosed);
-                        if (document.IsOpenInEditor)
-                        {
-                            _onOpened(document, EventArgs.Empty);
-                        }
-
-                        break;
-                    }
-
-                case ProjectChangeKind.DocumentRemoved:
+                await _foregroundDispatcher.RunOnForegroundAsync(async () =>
```

Comment:
You should already be on the fake foreground thread. ProjectManager_Changed events should alwyas be dispatched on it